src/freertos_drivers/esp32/Esp32Gpio.hxx

```diff
@@ -0,0 +1,467 @@
+            LOG(VERBOSE, "Esp32Gpio(%d) configured as OUTPUT", PIN_NUM);
+        }
+        else
+        {
+            ESP_ERROR_CHECK(gpio_set_direction(PIN_NUM, GPIO_MODE_INPUT));
+            LOG(VERBOSE, "Esp32Gpio(%d) configured as INPUT", PIN_NUM);
+        }
+    }
+
+    /// Gets the GPIO direction.
+    /// @return @ref DINPUT or @ref DOUTPUT
+    Direction direction() const override
+    {
+        if (IS_GPIO_OUTPUT(PIN_NUM))
+        {
+            return Direction::DOUTPUT;
+        }
+        return Direction::DINPUT;
+    }
+private:
+    template <class Defs, bool SAFE_VALUE, bool INVERT> friend struct GpioOutputPin;
+    template <class Defs, bool PUEN, bool PDEN> friend struct GpioInputPin;
+    /// Static instance variable that can be used for libraries expecting a
+    /// generic Gpio pointer. This instance variable will be initialized by the
+    /// linker and (assuming the application developer initialized the hardware
+    /// pins in hw_preinit) is accessible, including virtual methods at static
+    /// constructor time.
+    static const Esp32Gpio instance_;
+};
+
+/// Defines the linker symbol for the wrapped Gpio instance.
+template <gpio_num_t PIN_NUM, bool INVERTED>
+const Esp32Gpio<PIN_NUM, INVERTED> Esp32Gpio<PIN_NUM, INVERTED>::instance_;
+
+/// Parametric GPIO output class.
+/// @param Defs is the GPIO pin's definition base class, supplied by the
+/// GPIO_PIN macro.
+/// @param SAFE_VALUE is the initial value for the GPIO output pin.
+/// @param INVERT inverts the high/low state of the pin when set.
+template <class Defs, bool SAFE_VALUE, bool INVERT = false>
+struct GpioOutputPin : public Defs
+{
+public:
+    using Defs::PIN_NUM;
+// compile time sanity check that the selected pin is valid for output.
+#if defined(CONFIG_IDF_TARGET_ESP32S2) || defined(CONFIG_IDF_TARGET_ESP32S3)
+    static_assert(PIN_NUM != 46, "Pin 46 can not be used for output.");
+#else
+    static_assert(PIN_NUM < 34, "Pins 34 and above can not be used as output.");
+#endif // CONFIG_IDF_TARGET_ESP32S2 / CONFIG_IDF_TARGET_ESP32S3
+
+    /// Initializes the hardware pin.
+    static void hw_init()
+    {
+        gpio_pad_select_gpio(PIN_NUM);
+        ESP_ERROR_CHECK(gpio_reset_pin(PIN_NUM));
+        // using GPIO_MODE_INPUT_OUTPUT instead of GPIO_MODE_OUTPUT so that
+        // we can read the IO state
+        ESP_ERROR_CHECK(
+            gpio_set_direction(PIN_NUM, GPIO_MODE_INPUT_OUTPUT));
+        ESP_ERROR_CHECK(gpio_set_level(PIN_NUM, SAFE_VALUE));
+    }
+
+    /// Sets the hardware pin to a safe value.
+    static void hw_set_to_safe()
+    {
+        ESP_ERROR_CHECK(gpio_set_level(PIN_NUM, SAFE_VALUE));
+    }
+
+    /// Sets the output pin @param value if true, output is set to HIGH, if
+    /// false, output is set to LOW.
+    static void set(bool value)
+    {
+        if (INVERT)
+        {
+            ESP_ERROR_CHECK(gpio_set_level(PIN_NUM, !value));
```

Comment:
You have two implementations of GPIO here; one static and one with virtual functions and overrides.
This is not generally necessary, although some historic implementations may exist in two copies. Generally is is sufficient to have the one with static methods, then use GpioWrapper<>::instance() as the dynamic implementation.

I think the Stm32Gpio.hxx is a good example for simplification. I should make the TI and CC32xx follow the same pattern some day.

Reply:
I'll take a look at Stm32Gpio and see if I can rework this to be closer to that.

Reply:
I looked at switching to use `GpioWrapper<>::instance()` directly and it would pose a problem with using the instances with `MultiConfiguredPC`. The problem being `GpioWrapper::set_direction()` assuming that the underlying pin configuration can not be changed at runtime, that is true for some platforms but not necessarily so on the ESP32 (only certain pins have restrictions on them).

I'll look at removing the second implementation so that there is only one implementation (Esp32Gpio).